api/v1alpha1/util.go

```diff
@@ -270,3 +272,39 @@ func FleetSpecsAreEqual(f1, f2 FleetSpec) bool {
 
 	return DeviceSpecsAreEqual(f1.Template.Spec, f2.Template.Spec)
 }
+
+type MissingKeyGetOrError struct {
+	Key string
+}
+
+func (e *MissingKeyGetOrError) Error() string {
+	return fmt.Sprintf("key %s not defined", e.Key)
+}
+
+func GetGoTemplateFuncMap() template.FuncMap {
+	getOrDefault := func(m map[string]string, key string, defaultValue string) string {
+		if val, ok := m[key]; ok {
+			return val
+		}
+		return defaultValue
+	}
+
+	getOrError := func(m map[string]string, key string) (string, error) {
+		if val, ok := m[key]; ok {
+			return val, nil
+		}
+		return "", &MissingKeyGetOrError{Key: key}
+	}
+
+	replace := func(input, old, new string) string {
+		return strings.Replace(input, old, new, -1)
+	}
+
+	return template.FuncMap{
+		"toUpper":      strings.ToUpper,
+		"toLower":      strings.ToLower,
+		"replace":      replace,
+		"getOrDefault": getOrDefault,
+		"getOrError":   getOrError,
+	}
```

Comment:
Would it make sense to use the same [function names as Helm](https://helm.sh/docs/chart_template_guide/function_list/#string-functions) which adds [Sprig functions](https://masterminds.github.io/sprig/) to the built-in go-template functions?

--> "upper", "lower", "", "replace", "default", ??

I couldn't find an equivalent of "getOrError". I think one would usually use the `index` built-in function (`index .Metadata.Labels "key"`), which returns the empty value if the key doesn't exist, which can then be defaulted. And so if we were to set `missingkey=error`, we'd get the behaviour of `getOrError` if the `.` was used to access `.Metadata.Labels.key`). It'd be familiar to Helm users at least, but may be surprising to others?

Reply:
OK, so:
1. toUpper -> upper
2. toLower -> lower
3. Use missingkey=error and remove getOrError

> I think one would usually use the index built-in function (index .Metadata.Labels "key"), which returns the empty value if the key doesn't exist, which can then be defaulted.
What do you mean?  How can it "then be defaulted"?